src/goto-programs/loop_unroll.cpp

```diff
@@ -0,0 +1,213 @@
+#include <goto-programs/loop_unroll.h>
+
+bool unsound_loop_unroller::runOnLoop(loopst &loop, goto_programt &goto_program)
+{
+  int bound = get_loop_bounds(loop);
+  if(bound <= 0)
+    return false; // Can't unroll this
+
+  // Get loop exit goto number
+  goto_programt::targett loop_exit = loop.get_original_loop_exit();
+  // TODO: If we want a sound version: loop_exit++ so we have the skip instruction
+  if(loop_exit != goto_program.instructions.begin())
+  {
+    goto_programt::targett t_before = loop_exit;
+
+    if(t_before->is_goto() && is_true(t_before->guard))
+    {
+      // no 'fall-out'
+    }
+    else
+    {
+      // guard against 'fall-out'
+      goto_programt::targett t_goto = goto_program.insert(loop_exit);
+
+      t_goto->make_goto(loop_exit);
+      t_goto->location = loop_exit->location;
+      t_goto->function = loop_exit->function;
+      t_goto->guard = gen_true_expr();
+    }
+  }
+  /* TODO: For sound version
+  goto_programt::targett t_skip = goto_program.insert(loop_exit);
+  goto_programt::targett loop_iter = t_skip;
+
+  const goto_programt::targett loop_head = loop->get_original_loop_head();
+
+  t_skip->make_skip();
+  t_skip->location = loop_head->location;
+  t_skip->function = loop_head->function;
+  */
+
+  // If there is an inner control flow we need a map for it
+  std::map<goto_programt::targett, unsigned> target_map;
+  {
+    unsigned count = 0;
+    goto_programt::targett t = loop.get_original_loop_head();
+    t++; // get first instruction of the loop
+    for(; t != loop_exit; t++, count++)
+    {
+      assert(t != goto_program.instructions.end());
+      target_map[t] = count;
+    }
+  }
+
+  // re-direct any branches that go to loop_head to loop_iter
+  // TODO: not sure how to deal with this
```

Comment:
Weird comment. What do you need to know here?

Reply:
Removed this code snippet